Objects/bytearrayobject.c

```diff
@@ -267,6 +267,7 @@ PyByteArray_Concat(PyObject *a, PyObject *b)
     result = (PyByteArrayObject *) \
         PyByteArray_FromStringAndSize(NULL, va.len + vb.len);
     if (result != NULL) {
+        assert(result->ob_bytes != NULL);
```

Comment:
It seems like it *can* be NULL if PyByteArray_FromStringAndSize() is called with size=0: if va.len+vb.len=0 (if both strings are empty).

At least, I don't see any code path handling handling this case.

Reply:
Python code `bytearray() + bytearray()` triggers exactly this case. It seems like with the glibc, `memcpy(NULL, _PyByteArray_empty_string, 0)` doesn't crash. But as I far as I recall, it's an undefined behavior.

We should avoid calling memcpy() with n=0.

See:

* https://bugs.python.org/issue22605
* https://bugs.python.org/issue27570
* https://bugs.python.org/issue31347
* https://stackoverflow.com/questions/29844298/is-it-legal-to-call-memcpy-with-zero-length-on-a-pointer-just-past-the-end-of-an

Try also UBSan of clang.

cc @benjaminp @tiran

Reply:
Right, this `assert` is not correct.

What about only doing the first `memcpy` if `va.len` is non-zero, and the second one iff `vb.len` is non-zero? ISTM that if memcpy is inlined, the compiler would have a chance to skip the check.